### DISCUSSION TITLE
Possibility for separate threads for rendering DisplayList for each Window?

### DISCUSSION THREAD

**Discussion by 2jfw** [airsdk/Adobe-Runtime-Support#2515] [Q&A]
I cannot find the discussion where it was stated that AIR is using several threads for different tasks, like DisplayList Rendering etc.
@ajwfrost: My question is: would it be possible to have a separate rendering thread for the DisplayList for each AIR Native Window for better performance?

**Comment by ajwfrost** [comment]
Interesting idea, I had also wondered the same a few times recently! Currently, for a multi-window application, all of the windows (and therefore all of the stages/display lists and all the 'enter frame' handlers etc) are all in a single thread. Under the hood, AIR uses multiple threads to optimise the actual rendering of content, but each window still runs under a common thread in terms of message/event handling for the final blits (which is why there were odd impacts where minimising one window affected the performance of another..)

Trying to switch this though: I suspect we'd hit a problem within the ActionScript Virtual Machine whereby each thread has a separate memory management/garbage collection instance, and you can't then allocate/free memory from the wrong thread. So things like adding a display object to a different window (like you can do by allocating a StageWebView element to a different nativewindow's stage) would cause all sorts of problems, or trying to dispatch events from one window to another..

This is basically why, for Workers, there is no direct communication that's possible for sharing of ActionScript objects between two threads (each Worker is running in a different thread) - there are just a few synchronisation/communication objects and some common code on the C++ layer for ByteArrays (with a lot of special handling!)...

Hope that helps explain it! If you've got a specific use case that you're finding to be a bit of a bottleneck though, please let us know more details and we can see whether there are any more targeted optimisations that could be helpful.

thanks

**Comment by 2jfw** [comment]
Thanks for the details on this, Andrew!
Would the DisplayList rendering not be on top of this, meaning it just calculates the final output and thus could be in a separate rendering thread for each stage? I would think the rendering itself is kind of stateless and just reads and processes the data and creates the visual output  - but I'm likely missing points/details here so please excuse my naivety.

Regarding use cases, I would get back to you if we can point out a specific bottleneck.
In general, using Flex in a multi window multiple tabbed application, the object nesting can be quite deep and the display list processing becomes noticeably slower the more tabs and windows are open.
**EDIT:** There are normally several lists, trees, data grids and charts which are frequently updated so there is a heavy flow of incoming data which needs to be rendered. 
**EDIT2:** There are up to 5 AIR windows on 5 monitors open at the same time where each window likely has multiple tabbed subviews. [Self note: one task for us here would actually be to make sure that non-visible views/components should actually not be on the DisplayList]

**Comment by henke37** [comment]
The thing here is that it is a bidirectional dependency. Not only does threads have to render to a window (graphics apis are infamous for requiring thread affinity), but there is also the input processing. To use Windows as an example, all those window messages are only sent to the thread owning the window. It'd be a hassle to pass stuff between the thread owning the Window and the general actionscript thread, so they are the same thread.
